psh/test-kill.py

```diff
@@ -0,0 +1,94 @@
+# Phoenix-RTOS
+#
+# phoenix-rtos-tests
+#
+# psh kill command test
+#
+# Copyright 2021, 2022 Phoenix Systems
+# Author: Damian Loewnau
+#
+# This file is part of Phoenix-RTOS.
+#
+# %LICENSE%
+#
+
+import psh.tools.psh as psh
+from time import sleep
+
+
+def get_process_list(p):
+    ps_list = []
+    oneline_pattern = r'(.*?)(\r+)\n'
+    cmd = 'ps'
+    p.sendline(cmd)
+    p.expect(cmd + psh.EOL)
+    while True:
+        idx = p.expect([psh.PROMPT, oneline_pattern])
+        if idx == 0:
+            break
+        line = p.match.group(1)
+        try:
+            pid, ppid, pr, state, cpu, wait, time, vmem, thr, task = line.split(maxsplit=9)
+        except ValueError:
+            assert False, f'wrong ps output: {line}'
+
+        if pid.isdigit():
+            ps_list.append({'pid': pid, 'task': task, 'state': state})
+
+    return ps_list
+
+
+def create_psh_processes(p, count):
+    psh_pid_list = []
+    msg = 'Failed to create new psh process'
+    for _ in range(count):
+        psh.assert_exec(p, 'psh', '', msg)
+
+    # find pid list of sleep psh processes
+    for proc in get_process_list(p):
+        task = proc.get('task')
+        if proc.get('state') == 'sleep' and (task == 'psh' or task == '/bin/psh'):
+            psh_pid_list.append(proc.get('pid'))
+
+    psh_pid_count = len(psh_pid_list)
+    assert psh_pid_count == count, f'Created {psh_pid_count} psh processes, instead of {count}'
+
+    return psh_pid_list
+
+
+def assert_kill_procs(p, pid_list, pid_count):
+    for pid in pid_list:
+        msg = f'Wrong output when killing process with the following pid: {pid}'
+        psh.assert_cmd(p, f'kill {pid}', '', msg)
+    sleep(0.5)
+
+    for proc in get_process_list(p):
+        assert not proc['pid'] in pid_list, \
+            f'The process with id: {proc["pid"]} is still in the processes list, even though it has been killed'
+
+
+def assert_errors(p):
+    pid = 90
+    used_pid_list = [proc["pid"] for proc in get_process_list(p)]
+
+    # if process id is used, find next unused id
+    while str(pid) in used_pid_list:
+        pid = pid + 1
+
+    msg = 'Wrong kill help message'
+    psh.assert_cmd(p, 'kill', 'usage: kill <pid>', msg)
+    msg = 'Wrong kill error message when passing string in argument'
+    psh.assert_cmd(p, 'kill not_number', 'kill: could not parse process id: not_number', msg)
```

Comment:
I do not like checking exact error message output. Ideally should be done by checking return value but i assume this is currently not possible on psh.

Reply:
Good suggestion, there is some idea of creating additional psh command, which would return exit code. Once this is done, I will use it and possibly change my current tests.